flink-clients/src/main/java/org/apache/flink/client/deployment/application/DetachedApplicationRunner.java

```diff
@@ -78,7 +81,7 @@ public DetachedApplicationRunner(final boolean enforceSingleJobExecution) {
 			ClientUtils.executeProgram(executorServiceLoader, configuration, program, enforceSingleJobExecution, true);
 		} catch (ProgramInvocationException e) {
 			LOG.warn("Could not execute application: ", e);
-			throw new FlinkRuntimeException("Could not execute application.", e);
+			throw new CompletionException(new RestHandlerException("Could not execute application.", HttpResponseStatus.BAD_REQUEST, e));
```

Comment:
It is pretty weird that we throw the RestHandlerException here (along with a CompletionException with no CompletableFuture in sight); it has seemingly little to do with the REST layer.
I'd suggest to change the JarRunHandler to explicitly handle exceptions thrown by `applicationRunner.run` by changing the `thenApply` into a `handle`.